cpu/atmega2560/periph/uart.c

```diff
@@ -19,216 +19,115 @@
 {
-    /* initialize basic functionality */
-    int res = init_base(uart, baudrate);
-
-    if (res != 0) {
-        return res;
+    /* make sure the given device is valid */
+    if (uart >= UART_NUMOF) {
+        return -1;
     }
 
-    /* register callbacks */
-    config[uart].rx_cb = rx_cb;
-    config[uart].arg = arg;
-
-    /* configure interrupts and enable RX interrupt */
-    switch (uart) {
-#if UART_0_EN
-
-        case UART_0:
-            UART0_RX_IRQ_EN;
-            break;
-#endif /* UART_0_EN */
-#if UART_1_EN
-
-        case UART_1:
-            UART1_RX_IRQ_EN;
-            break;
-#endif /* UART_1_EN */
-#if UART_2_EN
-
-        case UART_2:
-            UART2_RX_IRQ_EN;
-            break;
-#endif /* UART_2_EN */
-#if UART_3_EN
-
-        case UART_3:
-            UART3_RX_IRQ_EN;
-            break;
-#endif /* UART_3_EN */
-    }
+    /* register interrupt context */
+    isr_ctx[uart].rx_cb = rx_cb;
+    isr_ctx[uart].arg   = arg;
 
-    return 0;
-}
+    /* disable and reset UART */
+    dev[uart]->CSRB = 0;
+    dev[uart]->CSRA = 0;
 
-static int init_base(uart_t uart, uint32_t baudrate)
-{
-    uint16_t clock_divider = CLOCK_CORECLOCK / (16 * baudrate);
-
-    switch (uart) {
-#if UART_0_EN
-
-        case UART_0:
-            /* enable RX and TX */
-            UART0_RX_TX_EN;
-            /* use 8 Bit characters */
-            UART0_SET_8BIT_SIZE;
-
-            /* set clock divider */
-            UART0_BAUD_RATE_L = clock_divider;
-            UART0_BAUD_RATE_H = (clock_divider >> 8);
-            break;
-#endif /* UART_0 */
-#if UART_1_EN
-
-        case UART_1:
-            /* enable RX and TX */
-            UART1_RX_TX_EN;
-            /* use 8 Bit characters */
-            UART1_SET_8BIT_SIZE;
-
-            /* set clock divider */
-            UART1_BAUD_RATE_L = clock_divider;
-            UART1_BAUD_RATE_H = (clock_divider >> 8);
-            break;
-#endif /* UART_1 */
-#if UART_2_EN
-
-        case UART_2:
-            /* enable RX and TX */
-            UART2_RX_TX_EN;
-            /* use 8 Bit characters */
-            UART2_SET_8BIT_SIZE;
-
-            /* set clock divider */
-            UART2_BAUD_RATE_L = clock_divider;
-            UART2_BAUD_RATE_H = (clock_divider >> 8);
-            break;
-#endif /* UART_2 */
-#if UART_3_EN
-
-        case UART_3:
-            /* enable RX and TX */
-            UART3_RX_TX_EN;
-            /* use 8 Bit characters */
-            UART3_SET_8BIT_SIZE;
-
-            /* set clock divider */
-            UART3_BAUD_RATE_L = clock_divider;
-            UART3_BAUD_RATE_H = (clock_divider >> 8);
-            break;
-#endif /* UART_3 */
-        default:
-            (void)clock_divider;    /* this makes cppcheck happy */
-    }
+    /* configure UART to 8N1 mode */
+    dev[uart]->CSRC = (1 << UCSZ00) | (1 << UCSZ01);
+    /* set clock divider */
+    dev[uart]->BRR = CLOCK_CORECLOCK / (16 * baudrate);
```

Comment:
Uhm... maybe I'm misunderstanding the manual but... it shouldn't be
```
dev[uart]->BRR = (CLOCK_CORECLOCK / 16 / baudrate - 1)
```

On the manual it says:
```
#define FOSC 1843200// Clock Speed
#define BAUD 9600
#define (MYUBRR FOSC/16/BAUD-1)
void main( void )
{...
USART_Init ( MYUBRR );
...} // main
void USART_Init( unsigned int ubrr){
/* Set baud rate */
UBRRH = (unsigned char)(ubrr>>8);
UBRRL = (unsigned char)ubrr;
/* Enable receiver and transmitter */
UCSRB = (1<<RXEN)|(1<<TXEN);
/* Set frame format: 8data, 2stop bit */
UCSRC = (1<<USBS)|(3<<UCSZ0);
} // USART_Init
```